### PR TITLE
Fbx animation

### PR DESCRIPTION
**Pull request by VTui22**

### CODE REVIEW

File: private/Producers/FbxProducer/FbxProducerImpl.cpp [hunk 1]
```diff
@@ -27,6 +27,35 @@ cd::Transform ConvertFbxNodeTransform(fbxsdk::FbxNode* pNode)
...
+
+	return cd::Transform(
+		cd::Vec3f(transformMatrix.GetTranslation()),
+		cd::Quaternion(cd::Quaternion::FromMatrix(transformMatrix.GetRotation())),
+		cd::Vec3f(transformMatrix.GetScale()));
+}
+
+fbxsdk::FbxAMatrix GetGeometryTransformation(fbxsdk::FbxNode* pNode)
+{
+	assert(pNode && "Null for mesh geometry");
```
Comment (T-rvw) [inline private/Producers/FbxProducer/FbxProducerImpl.cpp]: Don't use exception.

File: private/Producers/FbxProducer/FbxProducerImpl.cpp [hunk 2]
```diff
@@ -129,65 +185,43 @@ void FbxProducerImpl::Execute(cd::SceneDatabase* pSceneDatabase)
...
-		return cd::UpVector::YAxis;
-	};
-
-	int sign;
-	cd::FrontVector frontVector = fileAxisSystem.GetFrontVector(sign) == fbxsdk::FbxAxisSystem::eParityEven ? cd::FrontVector::ParityEven : cd::FrontVector::ParityOdd;
-	assert(sign == 1);
-
-	cd::AxisSystem axisSystem(handedness, GetUpVector(), frontVector);
-	pSceneDatabase->SetAxisSystem(cd::MoveTemp(axisSystem));
+		// Query file axis system and unit system.
```
Comment (T-rvw) [inline private/Producers/FbxProducer/FbxProducerImpl.cpp:194]: Doesn't apply AxisSystem so comment out all related codes.

File: private/Producers/FbxProducer/FbxProducerImpl.cpp [hunk 3]
```diff
@@ -480,21 +512,29 @@ cd::MeshID FbxProducerImpl::AddMesh(const fbxsdk::FbxMesh* pFbxMesh, const char*
...
+
 	// TODO : Multiple vertex color sets if necessary.
 	if (pLayerElementColorData)
 	{
 		mesh.SetVertexColorSetCount(1U);
 		meshVertexFormat.AddAttributeLayout(cd::VertexAttributeType::Color, cd::GetAttributeValueType<cd::Vec4f::ValueType>(), cd::Vec4f::Size);
 	}
 
+	int32_t skinDeformerCount = pFbxMesh->GetDeformerCount(fbxsdk::FbxDeformer::eSkin);
+	if (skinDeformerCount > 0)
```
Comment (T-rvw) [inline private/Producers/FbxProducer/FbxProducerImpl.cpp:537]: Need to check if skin deformer exists.

File: private/Producers/FbxProducer/FbxProducerImpl.cpp [hunk 4]
```diff
@@ -643,6 +684,86 @@ cd::MeshID FbxProducerImpl::AddMesh(const fbxsdk::FbxMesh* pFbxMesh, const char*
...
 
+	// Skin
+	std::map<uint32_t, uint32_t> mapVertexIndexToCurrentBoneCount;
+	for (int32_t deformerIndex = 0; deformerIndex < skinDeformerCount; ++deformerIndex)
+	{
+		auto* pSkinDeformer = static_cast<fbxsdk::FbxSkin*>(pFbxMesh->GetDeformer(deformerIndex, fbxsdk::FbxDeformer::eSkin));
+		pSceneDatabase->SetBoneCount(pSkinDeformer->GetClusterCount());
+		for (int32_t clusterIndex = 0; clusterIndex < pSkinDeformer->GetClusterCount(); ++clusterIndex)
+		{
+			// Cluster is an intermediate link node which connects bones to mesh shapes.
```
Comment (T-rvw) [inline private/Producers/FbxProducer/FbxProducerImpl.cpp:700]: SkinCluster is not a mesh node. See:
image

**Comment by T-rvw** [comment]
See [3c845228bf4993d1fe7321623b1592dd6d9a5ad0](https://github.com/CatDogEngine/AssetPipeline/pull/247/commits/3c845228bf4993d1fe7321623b1592dd6d9a5ad0). I reverted some changes maybe not reasonable. Please review it.

File: private/Producers/FbxProducer/FbxProducerImpl.cpp [hunk 5]
```diff
@@ -570,7 +558,7 @@ cd::MeshID FbxProducerImpl::AddMesh(const fbxsdk::FbxMesh* pFbxMesh, const char*
 	uint32_t currentPolygonCount = 0U;
 	std::map<uint32_t, uint32_t> mapVertexIDToControlPointIndex;
 	std::map<uint32_t, std::vector<uint32_t>> mapControlPointIndexToVertexID;
-	for (uint32_t polygonIndex = 0U; polygonIndex < availablePolygonCount; ++polygonIndex)
```
Comment (T-rvw) [inline private/Producers/FbxProducer/FbxProducerImpl.cpp]: This is a merge error in this PR. Need to revert!

File: private/Producers/FbxProducer/FbxProducerImpl.cpp [hunk 6]
```diff
@@ -598,10 +586,8 @@ cd::MeshID FbxProducerImpl::AddMesh(const fbxsdk::FbxMesh* pFbxMesh, const char*
 		polygon.reserve(polygonVertexCount);
 		for (uint32_t polygonVertexIndex = 0U; polygonVertexIndex < polygonVertexCount; ++polygonVertexIndex)
 		{
-			//fbxTransformMatrix fir axis transform.
-			fbxsdk::FbxAMatrix fbxTransformMatrix = pFbxMesh->GetNode()->EvaluateGlobalTransform();
 			uint32_t controlPointIndex = pFbxMesh->GetPolygonVertex(polygonIndex, polygonVertexIndex);
-			fbxsdk::FbxVector4 position = fbxTransformMatrix.MultT(pMeshVertexPositions[controlPointIndex]);
```
Comment (T-rvw) [inline private/Producers/FbxProducer/FbxProducerImpl.cpp]: Should not need to multiply mesh node transform. Please show me an example fbx file to see if you really need it.

File: private/Producers/FbxProducer/FbxProducerImpl.cpp [hunk 7]
```diff
@@ -632,10 +618,6 @@ cd::MeshID FbxProducerImpl::AddMesh(const fbxsdk::FbxMesh* pFbxMesh, const char*
 
 		if (applyTangentData)
 		{
-			if (!pLayerElementTangentData)
-			{
-				mesh.ComputeVertexTangents();
-			}
```
Comment (T-rvw) [inline private/Producers/FbxProducer/FbxProducerImpl.cpp]: Should not fill data in the produce step.

File: private/Scene/SceneDatabaseImpl.cpp [hunk 8]
```diff
@@ -409,8 +435,8 @@ void SceneDatabaseImpl::Validate() const
 	for (uint32_t trackIndex = 0U; trackIndex < GetTrackCount(); ++trackIndex)
 	{
 		const cd::Track& track = GetTrack(trackIndex);
-		//assert(trackIndex == track.GetID().Data());
-		//assert(track.GetTranslationKeyCount() > 0 || track.GetRotationKeyCount() > 0 || track.GetScaleKeyCount() > 0);
+		assert(trackIndex == track.GetID().Data());
+		assert(track.GetTranslationKeyCount() > 0 || track.GetRotationKeyCount() > 0 || track.GetScaleKeyCount() > 0);
```
Comment (T-rvw) [inline private/Scene/SceneDatabaseImpl.cpp]: Why you need to skip these checks?

File: private/Producers/FbxProducer/FbxProducerImpl.cpp (hunk 6 above)
Comment (VTui22) [reply private/Producers/FbxProducer/FbxProducerImpl.cpp]: image
I think this is a baking operation. When I change the coordinate system, the mesh does not correspond to the correct coordinate axes. After I perform this step, it matches correctly. I think if our node hierarchy is correct, this step is not be needed.

File: private/Producers/FbxProducer/FbxProducerImpl.cpp (hunk 4 above)
Comment (VTui22) [reply private/Producers/FbxProducer/FbxProducerImpl.cpp:700]: yes,my statement is not accurate. it's a corresponding relationship

File: private/Producers/FbxProducer/FbxProducerImpl.cpp [hunk 9]
```diff
@@ -49,7 +49,13 @@ cd::Transform ConvertFbxMatrixToTranform(fbxsdk::FbxAMatrix Matrix)
 
 fbxsdk::FbxAMatrix GetGeometryTransformation(fbxsdk::FbxNode* pNode)
 {
-	assert(pNode && "Null for mesh geometry");
+	if (!pNode)
+	{
+		std::printf("Null for mesh geometry");
```
Comment (T-rvw) [inline private/Producers/FbxProducer/FbxProducerImpl.cpp]: Should not use printf inside implementation unless it has an option such as DumpSceneDatabase.

File: private/Producers/FbxProducer/FbxProducerImpl.cpp (hunk 6 above)
Comment (T-rvw) [reply private/Producers/FbxProducer/FbxProducerImpl.cpp]: You can add an option to producer API.

File: private/Scene/SceneDatabaseImpl.cpp [hunk 10]
```diff
@@ -35,6 +35,13 @@ void Dump(const cd::Transform& transform)
 	details::Dump("\tScale", transform.GetScale());
 }
 
+void Dump(const cd::Matrix4x4& matrix)
+{
+	details::Dump("\tRestPoseTranslation", matrix.GetTranslation());
+	details::Dump("\tRestPoseRotation", cd::Quaternion::FromMatrix(matrix.GetRotation()));
+	details::Dump("\tRestPoseScale", matrix.GetScale());
+}
```
Comment (T-rvw) [inline private/Scene/SceneDatabaseImpl.cpp]: Not every matrix represents RestPostMatrix, please add a laber parameter such as other Dump functions.

**Pull request merged**